src/js/route.js

```diff
@@ -1,267 +1,218 @@
-import UrlBuilder from './UrlBuilder';
-import { stringify } from 'qs';
+import { parse, stringify } from 'qs';
 
-class Router extends String {
-    constructor(name, params, absolute, customZiggy = null) {
-        super();
+class Route {
+    absolute = true;
+    bindings = {};
 
-        this.name = name;
-        this.absolute = absolute;
-        this.ziggy = customZiggy ? customZiggy : Ziggy;
-        this.urlBuilder = this.name ? new UrlBuilder(name, absolute, this.ziggy) : null;
-        this.template = this.urlBuilder ? this.urlBuilder.construct() : '';
-        this.urlParams = this.normalizeParams(params);
-        this.queryParams = {};
-        this.hydrated = '';
+    constructor(name, definition, config) {
+        Object.entries({ name, ...definition, ...config }).map(([key, value]) => this[key] = value);
     }
 
-    normalizeParams(params) {
-        if (typeof params === 'undefined') return {};
+    // Get a template of the complete URL for this route, e.g. https://{team}.ziggy.dev/user/{user}
+    get template() {
+        // If  we're building a relative URL there's no origin, otherwise: if this route has a custom
+        // domain we construct the origin with that, and if not we use the app URL
+        const origin = !this.absolute ? '' : this.domain
+            ? `${this.baseProtocol}://${this.domain.replace(/\/$/, '')}${this.basePort ? `:${this.basePort}` : ''}`
+            : this.baseUrl.replace(/\/$/, '');
 
-        // If you passed in a string or integer, wrap it in an array
-        params = typeof params !== 'object' ? [params] : params;
-
-        this.numericParamIndices = Array.isArray(params);
-        return Object.assign({}, params);
+        return `${origin}/${this.uri}`.replace(/\/$/, '');
     }
 
-    with(params) {
-        this.urlParams = this.normalizeParams(params);
-        return this;
+    // Get an array of objects representing the parameters that this route accepts,
+    // e.g. [{ name: 'team', required: true }, { name: 'user', required: false }]
+    get segments() {
```

Comment:
If we could come up with a name that indicates "parameterizedSegementsButNotTheOtherOnes" that'd be awesome, but if not this is a good name